### PR TITLE
Refactoring of initial sync

### LINKED ISSUE #4815: Separate Block Fetching from Processing
In initial sync, currently block downloading and processing are tightly coupled together. However this leads to a penalty in our syncing speeds as we have to wait for other peers to return blocks back before being able to process them.

What we should be able to do is concurrently download blocks and also process them at the same time. What this will accomplish is allow us to seamlessly process one block after the next, while not being hampered by the response speeds of our peers.

### LINKED ISSUE #4969: Beacon client repeatedly processing slots during init sync
The latest beacon client (af5cc31565629f3e766b11caf7d8a7560d6c8901) processes each block lots of times before moving on to the next one during init sync:

```
[2020-02-29 22:10:36]  INFO initial-sync: Processing block 362641/366652 - estimated time remaining 8m21s blocksPerSecond=8.0 peers=13/17
[2020-02-29 22:10:36]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m18s blocksPerSecond=8.1 peers=13/17
[2020-02-29 22:10:36]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m15s blocksPerSecond=8.1 peers=13/17
[2020-02-29 22:10:36]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m34s blocksPerSecond=7.8 peers=13/17
[2020-02-29 22:10:36]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m31s blocksPerSecond=7.8 peers=13/17
[2020-02-29 22:10:36]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m27s blocksPerSecond=7.9 peers=13/17
[2020-02-29 22:10:36]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m24s blocksPerSecond=8.0 peers=13/17
[2020-02-29 22:10:36]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m21s blocksPerSecond=8.0 peers=13/17
[2020-02-29 22:10:36]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m18s blocksPerSecond=8.1 peers=13/17
[2020-02-29 22:10:37]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m15s blocksPerSecond=8.1 peers=13/17
[2020-02-29 22:10:37]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m12s blocksPerSecond=8.2 peers=13/17
[2020-02-29 22:10:37]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m34s blocksPerSecond=7.8 peers=13/17
[2020-02-29 22:10:37]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m31s blocksPerSecond=7.8 peers=13/17
[2020-02-29 22:10:37]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m27s blocksPerSecond=7.9 peers=13/17
[2020-02-29 22:10:37]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m24s blocksPerSecond=8.0 peers=13/17
[2020-02-29 22:10:37]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m21s blocksPerSecond=8.0 peers=13/17
[2020-02-29 22:10:37]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m18s blocksPerSecond=8.1 peers=13/17
[2020-02-29 22:10:38]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m15s blocksPerSecond=8.1 peers=13/17
[2020-02-29 22:10:38]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m12s blocksPerSecond=8.2 peers=13/17
[2020-02-29 22:10:38]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m31s blocksPerSecond=7.8 peers=13/17
[2020-02-29 22:10:38]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m27s blocksPerSecond=7.9 peers=13/17
[2020-02-29 22:10:38]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m24s blocksPerSecond=8.0 peers=13/17
[2020-02-29 22:10:38]  INFO initial-sync: Processing block 362641/366653 - estimated time remaining 8m21s blocksPerSecond=8.0 peers=13/17
[2020-02-29 22:10:38]  INFO initial-sync: Processing block 362642/366653 - estimated time remaining 8m18s blocksPerSecond=8.1 peers=13/17
[2020-02-29 22:10:39]  INFO initial-sync: Processing block 362642/366653 - estimated time remaining 8m40s blocksPerSecond=7.7 peers=13/17
[2020-02-29 22:10:39]  INFO initial-sync: Processing block 362642/366653 - estimated time remaining 8m37s blocksPerSecond=7.8 peers=13/17
[2020-02-29 22:10:39]  INFO initial-sync: Processing block 362642/366653 - estimated time remaining 8m34s blocksPerSecond=7.8 peers=13/17
[2020-02-29 22:10:39]  INFO initial-sync: Processing block 362642/366653 - estimated time remaining 8m30s blocksPerSecond=7.8 peers=13/17
[2020-02-29 22:10:40]  INFO initial-sync: Processing block 362642/366653 - estimated time remaining 8m27s blocksPerSecond=7.9 peers=13/17
[2020-02-29 22:10:40]  INFO initial-sync: Processing block 362642/366653 - estimated time remaining 8m24s blocksPerSecond=8.0 peers=13/17
[2020-02-29 22:10:40]  INFO initial-sync: Processing block 362642/366653 - estimated time remaining 8m51s blocksPerSecond=7.5 peers=13/17
[2020-02-29 22:10:40]  INFO initial-sync: Processing block 362642/366653 - estimated time remaining 8m47s blocksPerSecond=7.6 peers=13/17
```

This is unexpected and looks like a regression to me, earlier versions did not do this.

Edit: After restarting the client it continued syncing with the normal 1 log message per slot.

### PR DESCRIPTION
**Pull request by farazdagi**
- Resolves #4815
- Resolves #4969
- Related to #3898

--- 
Final PR for the refactored init-sync queue implementation

Features:
- all the previous code is moved into `beacon-chain/sync/initial-sync-v1/`, everything is gated by `--enable-initial-sync-queue`. If the flag is not provided, original implementation of service is used.
- updates `round_robin.go`

Please note that this PR line count is misleading (due to copying of previous implementation into `initial-sync-v1`). Actual delta is way more modest (updates to `round_robin.go`).

### CODE REVIEW

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 1]
```diff
@@ -74,16 +72,16 @@ func newBlocksFetcher(ctx context.Context, cfg *blocksFetcherConfig) *blocksFetc
...
 	return &blocksFetcher{
 		ctx:            ctx,
 		cancel:         cancel,
 		headFetcher:    cfg.headFetcher,
 		p2p:            cfg.p2p,
 		rateLimiter:    rateLimiter,
-		fetchRequests:  make(chan *fetchRequestParams, fetchRequestsBuffer),
-		fetchResponses: make(chan *fetchRequestResponse),
+		fetchRequests:  make(chan *fetchRequestParams, queueMaxPendingRequests),
+		fetchResponses: make(chan *fetchRequestResponse, queueMaxPendingRequests),
```
Comment (shayzluf) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go]:
```suggestion
		fetchResponses: make(chan *fetchRequestResponse, pendingQueueSize),
```
maybe a more general name if its not just the size of requests queue

File: beacon-chain/node/node.go [hunk 2]
```diff
@@ -404,9 +405,19 @@ func (b *BeaconNode) registerSyncService(ctx *cli.Context) error {
...
-		return err
+	var initSync prysmsync.Checker
+	if cfg := featureconfig.Get(); cfg.EnableInitSyncQueue {
+		var initSyncTmp *initialsync.Service
+		if err := b.services.FetchService(&initSyncTmp); err != nil {
+			return err
+		}
+		initSync = initSyncTmp
+	} else {
+		var initSyncTmp *initialsyncv1.Service
```
Comment (terencechain) [inline beacon-chain/node/node.go]: isn't v1 suppose to be newer? wouldnt that be the one that's gated?
perhaps a better name?

Comment (farazdagi) [reply beacon-chain/node/node.go]: I've assumed that I was working on v2 :) considered putting the old version under v0 tag, but supposed we have very solid working version not some prototype, so went with v1. Please let me know if you think that we need to change it.

File: beacon-chain/sync/initial-sync/blocks_queue.go [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+			}
+		}
+	}
+
+	return nil
+}
+
+// incrementCounter increments particular scheduler counter.
+func (s *schedulerState) incrementCounter(counter blockState, n uint64) {
+	s.Lock()
```
[the remark this answers is not in the document]
Comment (farazdagi) [reply beacon-chain/sync/initial-sync/blocks_queue.go]: This function is called on every processed block, it doesn't do much, doesn't take much (it just safely increments counter), so having overhead of the span doesn't seem like a good idea. Thoughts?

[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/sync/initial-sync/blocks_queue.go]: Ah ok. I agree with you

File: beacon-chain/node/node.go (hunk 2 above)
Comment (terencechain) [reply beacon-chain/node/node.go]: yeah I see the confusion since we don't put v2 for the new service, perhaps: 

s/initialsyncv1/initialsyncOld. ?

Comment (farazdagi) [reply beacon-chain/node/node.go]: Sounds good, will do it now.

File: beacon-chain/sync/initial-sync/blocks_queue.go [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+		}
+	}
+
+	return skippedBlocks, nil
+}
+
+// sendFetchedBlocks analyses available blocks, and sends them downstream in a correct slot order.
+// Blocks are checked starting from the current head slot, and up until next consecutive block is available.
+func (q *blocksQueue) sendFetchedBlocks(ctx context.Context) error {
+	q.state.sender.Lock()
```
[the remark this answers is not in the document]
Comment (farazdagi) [reply beacon-chain/sync/initial-sync/blocks_queue.go]: done!

File: beacon-chain/node/node.go (hunk 2 above)
Comment (farazdagi) [reply beacon-chain/node/node.go]: Done!

**Pull request merged**